### DISCUSSION TITLE
[Feature/Contribution] Allow adding a server SSL certificate (for Android phones)

### DISCUSSION THREAD

**Discussion by crisoagf** [immich-app/immich#13555] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

#### Summary

Hey 👋🏻 ,

I'd like to contribute a change to Immich adding the ability to import a trusted server certificate.

I have the code for what I believe to be a working solution, but before committing to writing tests I'd like to know if there's any interest in getting something like this merged.

In my mind, this would also remove the need for the blanket "Allow self-signed SSL certificates", for something slightly more secure. But that will not be in scope for this issue.

#### Description

First, thanks for Immich, I love it and it has been really helpful.

I have been really annoyed by the fact that I need to disable SSL certificate validation in Android to use my personal root certificate with it.

I am aware that the ideal solution (supporting User installed certificates in Android) is not possible (i.e. I have read #5657 ), but there's an alternate solution that fixes the problem for my use case (and I am guessing a lot of use cases, but I have no data to support this), which is just adding the ability to import a root certificate for Immich to use.

I already have developed this solution and it seems to work nicely: Just take what's already done for Client Authentication and slightly modify it to add a Server Root certificate.

Implementation is quite succint:
```
$ git diff --shortstat
 5 files changed, 67 insertions(+), 8 deletions(-)
```

Shall I write tests and open a PR?

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by alextran1502** [comment]
You can already import a cert in the App Settings > Advanced > scroll all the way down

**Reply by crisoagf** [reply]
You mean the one for Client Authentication? I am aware of that, that's not what I mean.

I mean adding a Server Certificate (or a Root certificate in general) to the trusted certificates, as an option for, e.g. accepting a specific self-signed server certificate without need to accept _all_ self-signed server certificates.

**Reply by bo0tzz** [reply]
The "allow self-signed SSL certificates" option.

**Reply by crisoagf** [reply]
That allows _any_ self-signed certificate, exposing me and my users to MITM attacks.

**Comment by crisoagf** [comment]
Hey @bo0tzz , I don't understand why this was closed, I have explained how it's different from the currently available options. This is sufficiently important to me for me to offer my time for this project, which I have stated in the description.

Can you at least explain what I am missing, and why the implementation that allows one to select a specific whitelisted certificate is completely replaceable by one that just accepts any certificate before just closing the Feature Request? Or should I open the PR first to show the code?

**Reply by alextran1502** [reply]
I think there will be some problems using a self-signed cert with Android since Android doesn't encourage that. You can test with your branch of playing a video remotely

**Reply by crisoagf** [reply]
That's a separate issue, also affecting users that tick the "Allow self-signed certificates" box.

For my use case that does not happen, because I am using my own Root cert that is imported to my Android's user cert store. I guess that should also work for Self-signed certificates?

**Reply by crisoagf** [reply]
Never mind that, turns out I was testing with 117, my branch has the same error, let me take a look to see if I can add that quickly.

**Reply by crisoagf** [reply]
I have added a comment to the video problems issue #5553 for what fixed video playback for me.

It doesn't change this feature request, I need both changes in order to use a specific certificate instead of allowing everything.

**Comment by crisoagf** [comment]
Hey hey, sorry coming back to this after so long!

I have opened PR #14335 , to demonstrate with code what I was aiming for. It still needs tests, but from my local validation (and from the similarity with client cert code), I am hopeful that it does what is written on the tin.

**Comment by mmomjian** [comment]
#15230
